### PR TITLE
fix: Can not use dynamic Service Account in Safer Cluster

### LINKED ISSUE #1524: Can not use dynamic Service Account in Safer Cluster
### TL;DR

Repetition of #27, but with safer-cluster module

### Terraform Resources

```markdown
# Create the nodeconfig service account
resource "google_service_account" "server" {
  account_id   = "server"
  display_name = "Server IAM"
  project      = var.project_id
}

module "gke" {
  source  = "terraform-google-modules/kubernetes-engine/google//modules/safer-cluster"
  version = "23.0.0"
  # some other stuff
  compute_engine_service_account = google_service_account.server.email
}
```

### Detailed design

```markdown
This is the same issue as #27 but on the safer-cluster module. 

│
│   on .terraform/modules/cortex-cluster.gke/modules/beta-private-cluster/sa.tf line 40, in resource "google_service_account" "cluster_service_account":
│   40:   count        = var.create_service_account ? 1 : 0

When compute_engine_service_account is dynamic, no determination can be made on this variable:
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/safer-cluster/main.tf#L114 until apply.

Passing create_service_account as default will fix this issue
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by davidcollom**
Fixes #1524

Where it is not possible to use the `safer-cluster` module without creating a service account from the upstream module

This leads to the following Error:

```
This is the same issue as #27 but on the safer-cluster module. 

│
│   on .terraform/modules/cortex-cluster.gke/modules/beta-private-cluster/sa.tf line 40, in resource "google_service_account" "cluster_service_account":
│   40:   count        = var.create_service_account ? 1 : 0

When compute_engine_service_account is dynamic, no determination can be made on this variable:
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/safer-cluster/main.tf#L114 until apply.

```

This fix took inspiration from #225

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
@davidcollom
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_generate_modules
 The modules need to be regenerated. Please run `make_build`. 
 ```bash 
Checking submodule's files generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/safer-cluster/README.md /tmp/tmp.tteNndozww/workspace/modules/safer-cluster/README.md
213a214
> | create\_compute\_engine\_service\_account | Defines if service account specified to run nodes should be created. | `bool` | `true` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/safer-cluster/main.tf /tmp/tmp.tteNndozww/workspace/modules/safer-cluster/main.tf
114c114
<   create_service_account = var.compute_engine_service_account == "" ? true : false
---
>   create_service_account = var.create_compute_engine_service_account
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/safer-cluster/variables.tf /tmp/tmp.tteNndozww/workspace/modules/safer-cluster/variables.tf
398a399,404
> variable "create_compute_engine_service_account" {
>   type        = bool
>   description = "Defines if service account specified to run nodes should be created."
>   default     = true
> }
> 
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/safer-cluster-update-variant/README.md /tmp/tmp.tteNndozww/workspace/modules/safer-cluster-update-variant/README.md
213a214
> | create\_compute\_engine\_service\_account | Defines if service account specified to run nodes should be created. | `bool` | `true` | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/safer-cluster-update-variant/main.tf /tmp/tmp.tteNndozww/workspace/modules/safer-cluster-update-variant/main.tf
114c114
<   create_service_account = var.compute_engine_service_account == "" ? true : false
---
>   create_service_account = var.create_compute_engine_service_account
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/safer-cluster-update-variant/variables.tf /tmp/tmp.tteNndozww/workspace/modules/safer-cluster-update-variant/variables.tf
398a399,404
> variable "create_compute_engine_service_account" {
>   type        = bool
>   description = "Defines if service account specified to run nodes should be created."
>   default     = true
> }
> 
Error: submodule's files generation has not been run, please run the
'make build' command and commit changes
 ```

**Comment by comment-bot-dev** [comment]
@davidcollom
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@davidcollom
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate .
Success!
 The configuration is valid.
terraform_validate ./examples/acm-terraform-blog-part1/terraform
Success!
 The configuration is valid.
terraform_validate ./examples/acm-terraform-blog-part2/terraform
Success!
 The configuration is valid.
terraform_validate ./examples/acm-terraform-blog-part3/terraform
Success!
 The configuration is valid.
terraform_validate ./examples/deploy_service
Success!
 The configuration is valid.
terraform_validate ./examples/disable_client_cert
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool_update_variant
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool_update_variant_beta
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool_update_variant_public_beta
Success!
 The configuration is valid.
terraform_validate ./examples/private_zonal_with_networking
Success!
 The configuration is valid.
terraform_validate ./examples/regional_private_node_pool_oauth_scopes
Success!
 The configuration is valid.
terraform_validate ./examples/safer_cluster
Success!
 The configuration is valid.
terraform_validate ./examples/safer_cluster_iap_bastion
Success!
 The configuration is valid.
terraform_validate ./examples/shared_vpc
Success!
 The configuration is valid.
terraform_validate ./examples/simple_autopilot_private
Success!
 The configuration is valid.
terraform_validate ./examples/simple_autopilot_private_non_default_sa
Success!
 The configuration is valid.
terraform_validate ./examples/simple_autopilot_public
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_beta
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_private
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_private_beta
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_with_gateway_api
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_with_kubeconfig
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_with_networking
Success!
 The configuration is valid.
terraform_validate ./examples/simple_windows_node_pool
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_private
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_acm
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_asm
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_hub
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_hub_kubeconfig
Success!
 The configuration is valid.
terraform_validate ./examples/stub_domains
Success!
 The configuration is valid.
terraform_validate ./examples/stub_domains_private
Success!
 The configuration is valid.
terraform_validate ./examples/stub_domains_upstream_nameservers
Success!
 The configuration is valid.
terraform_validate ./examples/upstream_nameservers
Success!
 The configuration is valid.
terraform_validate ./examples/workload_identity
Success!
 The configuration is valid.
terraform_validate ./examples/workload_metadata_config
Success!
 The configuration is valid.
terraform_validate ./modules/acm
Success!
 The configuration is valid.
terraform_validate ./modules/asm
Success!
 The configuration is valid.
terraform_validate ./modules/auth
Success!
 The configuration is valid.
terraform_validate ./modules/beta-autopilot-private-cluster
Success!
 The configuration is valid.
terraform_validate ./modules/beta-autopilot-public-cluster
Success!
 The configuration is valid.
terraform_validate ./modules/beta-private-cluster
Success!
 The configuration is valid.
terraform_validate ./modules/beta-private-cluster-update-variant
Success!
 The configuration is valid.
terraform_validate ./modules/beta-public-cluster
Success!
 The configuration is valid.
terraform_validate ./modules/beta-public-cluster-update-variant
Success!
 The configuration is valid.
terraform_validate ./modules/binary-authorization
╷
│
 
Error: 
Failed to query available provider packages
│
 
│
 
Could not retrieve the list of available versions for provider
│
 
hashicorp/google-beta: could not query provider registry for
│
 
registry.terraform.io/hashicorp/google-beta: read tcp
│
 
192.168.10.2:44950->99.84.160.22:443: read: connection reset by peer
╵
terraform_validate ./modules/fleet-membership
Success!
 The configuration is valid.
terraform_validate ./modules/hub-legacy
Success!
 The configuration is valid.
terraform_validate ./modules/private-cluster
Success!
 The configuration is valid.
terraform_validate ./modules/private-cluster-update-variant
Success!
 The configuration is valid.
terraform_validate ./modules/safer-cluster
Success!
 The configuration is valid.
terraform_validate ./modules/safer-cluster-update-variant
Success!
 The configuration is valid.
terraform_validate ./modules/services
Success!
 The configuration is valid.
terraform_validate ./modules/workload-identity
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/beta_cluster
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/deploy_service
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/disable_client_cert
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/node_pool
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/node_pool_update_variant
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/private_zonal_with_networking
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/safer_cluster
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/safer_cluster_iap_bastion
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/sandbox_enabled
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_autopilot_private
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_autopilot_public
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional_private
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional_with_gateway_api
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional_with_kubeconfig
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional_with_networking
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_windows_node_pool
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_zonal
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_zonal_private
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_zonal_with_asm
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/stub_domains
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/stub_domains_private
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/stub_domains_upstream_nameservers
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/upstream_nameservers
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/workload_identity
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/workload_metadata_config
Success!
 The configuration is valid.
terraform_validate ./test/setup
Success!
 The configuration is valid.
 ```

File: autogen/safer-cluster/main.tf.tmpl [hunk 1]
```diff
@@ -116,7 +116,7 @@ module "gke" {
   //   All applications should run with an identity defined via Workload Identity anyway.
   // - Use a service account passed as a parameter to the module, in case the user
   //   wants to maintain control of their service accounts.
-  create_service_account = var.compute_engine_service_account == "" ? true : false
+  create_service_account = var.create_compute_engine_service_account
```
Comment (ericyz) [inline autogen/safer-cluster/main.tf.tmpl:119]: I suggest to rename it to `create_service_account` to keep it consistency

Comment (ericyz) [reply autogen/safer-cluster/main.tf.tmpl:119]: The compute engine service account might not be accurate here

**Review by ericyz — CHANGES_REQUESTED** [review]
Minor comments.

**Pull request closed without merging**